File: scala/prime-factors/src/main/scala/PrimeFactors.scala
import scala.annotation.tailrec

object PrimeFactors {

  // needed for the tests to compile and run
  def apply() = this

  type Factors = List[Int]

  def primeFactors(number: Long): Factors = {
    unfoldRight((number, 2))(findNext)
  }

  private type Division = (Long, Int)

  @tailrec
  private def findNext(division: Division): Option[(Int,Division)] = {
    val (dividend, divisor) = division
    if (dividend == 1) None
    else if (divisor goesCleanlyInto dividend) Some(divisor, (dividend / divisor, divisor))
    else findNext((dividend, divisor + 1))
  }

  implicit class IntOps(divisor: Int) {
    def goesCleanlyInto(dividend: Long): Boolean =
      dividend % divisor == 0
  }

  private def unfoldRight[A,B](seed: B)(f: B => Option[(A,B)]): List[A] =
    f(seed) match {
    case None => Nil
    case Some((a,b)) => a :: unfoldRight(b)(f)
  }
}

File: haskell/prime-factors/PrimeFactors.hs
module PrimeFactors (primeFactors) where

import Data.List (unfoldr)

type PrimeFactors = [Int]

primeFactors :: Int -> PrimeFactors
primeFactors n =
  unfoldr findNext (n, 2)
  where
    findNext :: (Int, Int) -> Maybe (Int, (Int,Int))
    findNext (dividend, divisor)
      | dividend == 1 = Nothing
      | divisor `goesCleanlyInto` dividend = Just (divisor, (dividend `quot` divisor, divisor))
      | otherwise = findNext (dividend, divisor + 1)

    goesCleanlyInto :: Int -> Int -> Bool
    goesCleanlyInto divisor dividend =
      dividend `mod` divisor == 0
